### PR TITLE
Enable LMCache for cpuoffloading, LMCache docker support, enable lmcache

### PR DESCRIPTION
**Pull request by hsubramony**

### CODE REVIEW

File: vllm/v1/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -893,24 +905,25 @@ def _get_prompts_and_decodes(
         assert total_num_scheduled_tokens > 0
         num_reqs = self.input_batch.num_reqs
         assert num_reqs > 0
+        is_decoder_only = self.is_decoder_only()
 
         # Traverse decodes first
         decode_req_ids = []
+        #logger.info(f"rank {os.getenv('RANK')}, {scheduler_output}")
```
Comment (libinta) [inline vllm/v1/worker/hpu_model_runner.py:912]: remove

File: vllm/v1/serial_utils.py [hunk 2]
```diff
@@ -280,4 +280,8 @@ def ext_hook(self, code: int, data: memoryview) -> Any:
                 return cloudpickle.loads(data)
 
         raise NotImplementedError(
+<<<<<<< HEAD
             f"Extension type code {code} is not supported")
+=======
+            f"Extension type code {code} is not supported")
+>>>>>>> 77710b186 (Added PR17427, 16866, 16432, 16492,13790)
```
[the remark this answers is not in the document]
Comment (hsubramony) [reply vllm/v1/serial_utils.py]: done

File: vllm/v1/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -460,6 +462,7 @@ def forward(self, *args, **kwargs):
         attn_meta = kwargs.pop('attn_metadata')
         if 'kv_caches' in kwargs:
             kwargs.pop('kv_caches')
+        #import remote_pdb;remote_pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (hsubramony) [reply vllm/v1/worker/hpu_model_runner.py]: done

File: vllm/v1/worker/hpu_model_runner.py [hunk 4]
```diff
@@ -1094,7 +1107,7 @@ def _prepare_prefill_inputs(self,
         prefill_logits_indices = []
         block_table_cpu_tensor = self.input_batch.block_table.get_cpu_tensor()
         fake_prefix_prefill = False
-
+        #import remote_pdb;remote_pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (hsubramony) [reply vllm/v1/worker/hpu_model_runner.py]: done

File: vllm/attention/layer.py [hunk 5]
```diff
@@ -143,7 +143,7 @@ def __init__(
         # opaque custom op. For other platforms, we directly call them
         # and let torch.compile handle them.
         self.use_direct_call = not current_platform.is_cuda_alike(
-        ) and not current_platform.is_cpu()
+        ) and not current_platform.is_cpu() and not torch.hpu.is_available()
```
Comment (michalkuligowski) [inline vllm/attention/layer.py:146]: PLease fix mypy precommit errrors

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/attention/layer.py
Comment (liujianhang-design) [inline vllm/attention/layer.py:1]: No need to change this file. And Please remove all "diff" caused by format change.

File: vllm/distributed/kv_transfer/kv_connector/factory.py
Comment (liujianhang-design) [inline vllm/distributed/kv_transfer/kv_connector/factory.py:1]: I personally think we don't need add this kind of protection (check connector in supported list) by ourselves. We should NOT touch this kind of core file.

File: vllm/v1/core/sched/output.py
Comment (liujianhang-design) [inline vllm/v1/core/sched/output.py:1]: No need to change this file just for import something without using it.

File: vllm/v1/core/sched/scheduler.py
Comment (liujianhang-design) [inline vllm/v1/core/sched/scheduler.py:1]: No need to change this file just for import something without using it.

File: vllm/attention/layer.py
Comment (hsubramony) [reply vllm/attention/layer.py:1]: done

File: vllm/distributed/kv_transfer/kv_connector/factory.py
Comment (hsubramony) [reply vllm/distributed/kv_transfer/kv_connector/factory.py:1]: done

File: vllm/v1/core/sched/output.py
Comment (hsubramony) [reply vllm/v1/core/sched/output.py:1]: done

File: vllm/v1/worker/hpu_worker.py [hunk 6]
```diff
@@ -14,7 +14,8 @@
 import vllm.envs as envs
 from vllm.config import ParallelConfig, VllmConfig
 from vllm.distributed import (ensure_model_parallel_initialized,
-                              init_distributed_environment)
+                      get_pp_group, init_distributed_environment)
```
Comment (michalkuligowski) [inline vllm/v1/worker/hpu_worker.py:17]: Please fix precommit errors

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/v1/worker/hpu_model_runner.py [hunk 7]
```diff
@@ -492,7 +494,7 @@ def trim_attn_metadata(metadata: HPUAttentionMetadataV1) -> object:
     attention_metadata = subtuple(metadata, 'TrimmedAttentionMetadata', [
         'attn_bias', 'seq_lens_tensor', 'context_lens_tensor', 'block_list',
         'block_mapping', 'block_usage', 'slot_mapping', 'is_prompt',
-        'block_size', 'block_groups'
+        'block_size', 'block_groups', 'is_warmup'
```
Comment (madamczyk-intel) [inline vllm/v1/worker/hpu_model_runner.py:497]: This breaks warmup. HPU graphs are cached based on the hash of their inputs. If we warmup with 'is_warmup=True' then we will warmup different recipes then those that might be used later on during regular runs when 'is_warmup=False'

**Review by madamczyk-intel — CHANGES_REQUESTED** [review]

File: vllm/v1/worker/hpu_model_runner.py (hunk 7 above)
Comment (libinta) [reply vllm/v1/worker/hpu_model_runner.py]: removed

**Review by michalkuligowski — APPROVED** [review]

File: examples/others/lmcache/hpu/disagg_prefill_lmcache_v1/disagg_vllm_launcher.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    #UCX_TLS=tcp \
+    LMCACHE_CONFIG_FILE=$decode_config_file \
+        VLLM_ENABLE_V1_MULTIPROCESSING=1 \
+        VLLM_WORKER_MULTIPROC_METHOD=spawn \
+        LMCACHE_CHUNK_SIZE=256 \
+        RANK=1 \
+        DECODER_RANK=1 \
+        vllm serve $MODEL \
+        --port 1200 \
+        --gpu_memory_utilization 0.5 \
```
Comment (libinta) [inline examples/others/lmcache/hpu/disagg_prefill_lmcache_v1/disagg_vllm_launcher.sh:51]: remove this

**Review by xuechendi — APPROVED** [review]

File: vllm/v1/worker/hpu_model_runner.py [hunk 9]
```diff
@@ -404,10 +412,20 @@ def forward(self, *args, **kwargs):
         if 'kv_caches' in kwargs:
             kwargs.pop('kv_caches')
         with set_forward_context(attn_meta, self.vllm_config):
+            if not is_warmup:
```
Comment (xuechendi) [inline vllm/v1/worker/hpu_model_runner.py:415]: how does is_warmup relate to maybe_setup_kv_connector?
Does is_warmup as an indicator saying the kv_connector is already setup?
I personally think the name is_warmup can't self-explained here.

Can you add comments to explain when maybe_setup_kv_connector, maybe_setup_kv_connector supposed to be called.

File: vllm/forward_context.py [hunk 10]
```diff
@@ -65,7 +65,8 @@ def set_forward_context(attn_metadata: Any,
                         vllm_config: VllmConfig,
                         virtual_engine: int = 0,
                         num_tokens: int = 0,
-                        dp_awared_padding: bool = False):
+                        dp_awared_padding: bool = False,
+                        is_warmup: bool = False):
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/forward_context.py:121]: agrre with @michalkuligowski , since `set_forward_context` is general for all platform, I don't think we should add new parameter here.
Personally, I think using hpu_attn_metadata is better option.
Furthurmore, it will be also great to add description to 'is_warmup' on the condition and purpose. I don't think the term is self-explained here.

File: vllm/v1/worker/hpu_model_runner.py (hunk 9 above)
Comment (xuechendi) [reply vllm/v1/worker/hpu_model_runner.py:415]: referring to GPU impl, they don't use is_warmup as condition:
https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/gpu_model_runner.py#L1365-L1389

What is the purpose here? If that is something will fix later, please also add #TODO/#Fixme comments

File: examples/others/lmcache/hpu/README.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+This folder demonstrates how to use LMCache for disaggregated prefilling, CPU offloading and KV cache sharing.
+
+## 1. Disaggregated Prefill in vLLM v1
+
+This example demonstrates how to run LMCache with disaggregated prefill using lm or redis on a single node.
+
+### Prerequisites
+- HPU implementation to be upstreamed
+- At least 2 HPUs
```
Comment (xuechendi) [inline examples/others/lmcache/hpu/README.md:12]: cards or nodes?

File: examples/others/lmcache/hpu/README.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+to run disaggregated prefill and benchmark the performance.
+
+lmserver is default and it's configurable as well as tensor_parallel_size and model name.
+
+Example) redis server, tensor_parallel_size 4 and Llama-3.1-70B-Instruct model
+
+```
+PT_HPU_GPU_MIGRATION=1 VLLM_USE_V1=1 VLLM_SKIP_WARMUP=True PT_HPU_ENABLE_LAZY_COLLECTIVES=true bash disagg_example.sh -s redis -t 4 -m meta-llama/Llama-3.1-70B-Instruct
```
Comment (xuechendi) [inline examples/others/lmcache/hpu/README.md:32]: is it necessary to use PT_HPU_GPU_MIGRATION=1?
Will have bring performance impact?

File: examples/others/lmcache/hpu/README.md (hunk 11 above)
Comment (yeonsily) [reply examples/others/lmcache/hpu/README.md:12]: cards. will update it.

File: examples/others/lmcache/hpu/README.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+Please Note: HPU integration for LMCache will be upstreamed. After that, the following test cases can be used.
+
+This folder demonstrates how to use LMCache for disaggregated prefilling, CPU offloading and KV cache sharing.
+
+## 1. Disaggregated Prefill in vLLM v1
+
+This example demonstrates how to run LMCache with disaggregated prefill using lm or redis on a single node.
+
+### Prerequisites
+- HPU implementation to be upstreamed
```
[the remark this answers is not in the document]
Comment (yeonsily) [reply examples/others/lmcache/hpu/README.md:11]: should be removed.

File: examples/others/lmcache/hpu/README.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+- At least 2 HPUs
+- Valid Hugging Face token (HF_TOKEN) for Llama 3.1 8B Instruct.
+
```
[the remark this answers is not in the document]
Comment (yeonsily) [reply examples/others/lmcache/hpu/README.md:14]: Actually we don't have LMCache upstream PR yet. It's still in intel internal link.

**Review by xuechendi — APPROVED** [review]
